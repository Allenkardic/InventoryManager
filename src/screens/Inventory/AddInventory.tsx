import React from 'react';
import {ScrollView, StyleSheet, TextStyle, View} from 'react-native';
// components
import NavigationHeader from '../../components/NavigationHeader';
import CustomInput from '../../components/CustomInput';
import CustomButton from '../../components/CustomButton';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {StackModels} from '../../navigations/models';
// context
import {AuthUserContext} from '../../context/AuthUserContext';
import {InventoryListContext} from '../../context/InventoryContext';

import {
  stampId,
  checkIfItemNameExist,
  checkWordsCount,
} from '../../utils/constants';
import {COLORS, SPACING} from '../../utils/themes';

import {showMessage} from 'react-native-flash-message';
// form
import {useForm, Controller} from 'react-hook-form';
import {yupResolver} from '@hookform/resolvers/yup';
import * as yup from 'yup';

type Style = {
  formContainer: TextStyle;
  btnContainer: TextStyle;
};
// navigation type
type Props = NativeStackScreenProps<StackModels, 'AddInventory'>;

// form type
type FormValues = {
  inventoryName: string;
  stockCount: number;
  description: string;
  price: number;
};

function AddInventory({navigation}: Props) {
  const {authUserStore} = React.useContext(AuthUserContext);
  const {inventoryStore, setInventoryStore} =
    React.useContext(InventoryListContext);

  // filter inventory base on the user that created it
  const inventoryByUser = inventoryStore.filter(
    item => item.userEmail === authUserStore.email,
  );

  // validation schema
  const schema = yup.object().shape({
    inventoryName: yup
      .string()
      .required('Inventory name is required')
      .test(
        'inventoryName',
        'Inventory name already exists',
        (value: any) => checkIfItemNameExist(value, inventoryByUser) === false,
      ),
    stockCount: yup.number().required('Stock count left is required'),
    description: yup
      .string()
      .required('Brief descripting is required')
      .test(
        'description',
        'Description must not be less than 3 words',
        (value: any) => checkWordsCount(value) >= 3,
      ),
    price: yup.number().required('Price is required'),
  });

  // React Hook Form
  const {
    control,
    handleSubmit,
    formState: {errors},
  } = useForm<FormValues>({
    resolver: yupResolver(schema),
  });

  const onSubmit = (data: FormValues) => {
    const payload = {
      id: stampId(),
      userEmail: authUserStore.email,
      inventoryName: data.inventoryName,
      stockCount: data.stockCount,
      description: data.description,
      price: data.price,
    };

    const addNewPayload = [...inventoryStore, payload];
    // add data to inventory list
    setInventoryStore(addNewPayload);
    showMessage({
      type: 'success',
      message: 'Item added successfuly',
    });
    navigation.navigate('Inventory');
  };

  // console.log(checkIfItemNameExist('allen', inventoryStore), 'storesss');
  return (
    <>
      <View>
        <NavigationHeader
          onPressBackBtn={() => navigation.goBack()}
          title={'Add Inventory'}
          backgroundColor={COLORS.smokeWhite}
        />
      </View>
      <ScrollView>
        <View
          style={{
            ...styles.formContainer,
          }}>
          <Controller
            control={control}
            render={({field: {onChange, onBlur, value}}) => (
              <CustomInput
                onChangeText={onChange}
                value={value}
                placeholder="ex: book"
                label="Item name"
                backgroundColor={COLORS.lightWhite}
                error={errors.inventoryName?.message}
                marginBottom={SPACING.xsmall}
              />
            )}
            name="inventoryName"
          />
          <Controller
            control={control}
            render={({field: {onChange, onBlur, value}}) => (
              <CustomInput
                onChangeText={onChange}
                value={value}
                placeholder="ex: 200"
                label="Price"
                backgroundColor={COLORS.lightWhite}
                error={errors.price?.message}
                marginBottom={SPACING.xsmall}
                keyboardType={'number-pad'}
              />
            )}
            name="price"
          />

          <Controller
            control={control}
            render={({field: {onChange, onBlur, value}}) => (
              <CustomInput
                onChangeText={onChange}
                value={value}
                placeholder="ex: 10"
                label="Count"
                backgroundColor={COLORS.lightWhite}
                error={errors.stockCount?.message}
                marginBottom={SPACING.xsmall}
                keyboardType={'number-pad'}
              />
            )}
            name="stockCount"
          />

          <Controller
            control={control}
            render={({field: {onChange, onBlur, value}}) => (
              <CustomInput
                onChangeText={onChange}
                value={value}
                placeholder="A brief description"
                label="Description"
                multiline={true}
                backgroundColor={COLORS.lightWhite}
                error={errors.description?.message}
              />
            )}
            name="description"
          />
        </View>
      </ScrollView>
      <View
        style={{
          ...styles.btnContainer,
        }}>
        <CustomButton title="Submit" onPress={handleSubmit(onSubmit)} />
      </View>
    </>
  );
}

const styles = StyleSheet.create<Style>({
  formContainer: {
    paddingTop: SPACING.small,
    paddingHorizontal: SPACING.xsmall,
  },
  btnContainer: {
    paddingBottom: SPACING.xsmall,
    paddingHorizontal: SPACING.xsmall,
  },
});

export default AddInventory;
