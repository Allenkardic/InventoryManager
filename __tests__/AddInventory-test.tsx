/**
 * @format
 */

import 'react-native';
import React from 'react';
import AddInventory from '../src/screens/Inventory/AddInventory';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const createTestProps = (props: Object) => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...props,
});

describe('AddInventory Screen', () => {
  it('renders correctly', () => {
    let props: any;
    props: createTestProps;
    const tree = renderer.create(<AddInventory {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
