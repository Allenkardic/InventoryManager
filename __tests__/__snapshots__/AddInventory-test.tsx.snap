// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddInventory Screen renders correctly 1`] = `
Array [
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f5f5f5",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 12,
          "paddingTop": 96,
        }
      }
    >
      <View
        style={
          Object {
            "alignSelf": "flex-start",
            "width": "10%",
          }
        }
      >
        <Text
          allowFontScaling={false}
          onPress={[Function]}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#121515",
                "fontSize": 40,
              },
              undefined,
              Object {
                "fontFamily": "Feather",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          ÔÑê
        </Text>
      </View>
      <View
        style={
          Object {
            "width": "80%",
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Object {
              "alignContent": "center",
              "alignItems": "center",
              "fontSize": 31,
              "justifyContent": "center",
              "textAlign": "center",
            }
          }
        >
          Add Inventory
        </Text>
      </View>
      <View
        style={
          Object {
            "width": "10%",
          }
        }
      />
    </View>
  </View>,
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "paddingHorizontal": 48,
            "paddingTop": 64,
          }
        }
      >
        <View
          style={
            Object {
              "marginBottom": 48,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Object {
                "color": "#0A0A0A80",
                "fontSize": 20,
                "marginBottom": 4,
                "paddingLeft": 12,
              }
            }
          >
            Item name
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#fafafa",
                "borderColor": "#DDDDDD",
                "borderRadius": 10,
                "borderWidth": 1,
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "width": "100%",
              }
            }
          >
            <TextInput
              autoCapitalize="none"
              autoCorrect={false}
              multiline={false}
              numberOfLines={4}
              onChangeText={[Function]}
              placeholder="ex: book"
              placeholderTextColor="#0A0A0A80"
              selectTextOnFocus={true}
              style={
                Array [
                  Object {
                    "height": 66.5,
                    "lineHeight": 0,
                    "paddingLeft": 24,
                    "width": "100%",
                  },
                ]
              }
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 48,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Object {
                "color": "#0A0A0A80",
                "fontSize": 20,
                "marginBottom": 4,
                "paddingLeft": 12,
              }
            }
          >
            Price
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#fafafa",
                "borderColor": "#DDDDDD",
                "borderRadius": 10,
                "borderWidth": 1,
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "width": "100%",
              }
            }
          >
            <TextInput
              autoCapitalize="none"
              autoCorrect={false}
              keyboardType="number-pad"
              multiline={false}
              numberOfLines={4}
              onChangeText={[Function]}
              placeholder="ex: 200"
              placeholderTextColor="#0A0A0A80"
              selectTextOnFocus={true}
              style={
                Array [
                  Object {
                    "height": 66.5,
                    "lineHeight": 0,
                    "paddingLeft": 24,
                    "width": "100%",
                  },
                ]
              }
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 48,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Object {
                "color": "#0A0A0A80",
                "fontSize": 20,
                "marginBottom": 4,
                "paddingLeft": 12,
              }
            }
          >
            Count
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#fafafa",
                "borderColor": "#DDDDDD",
                "borderRadius": 10,
                "borderWidth": 1,
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "width": "100%",
              }
            }
          >
            <TextInput
              autoCapitalize="none"
              autoCorrect={false}
              keyboardType="number-pad"
              multiline={false}
              numberOfLines={4}
              onChangeText={[Function]}
              placeholder="ex: 10"
              placeholderTextColor="#0A0A0A80"
              selectTextOnFocus={true}
              style={
                Array [
                  Object {
                    "height": 66.5,
                    "lineHeight": 0,
                    "paddingLeft": 24,
                    "width": "100%",
                  },
                ]
              }
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 24,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Object {
                "color": "#0A0A0A80",
                "fontSize": 20,
                "marginBottom": 4,
                "paddingLeft": 12,
              }
            }
          >
            Description
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#fafafa",
                "borderColor": "#DDDDDD",
                "borderRadius": 10,
                "borderWidth": 1,
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "width": "100%",
              }
            }
          >
            <TextInput
              autoCapitalize="none"
              autoCorrect={false}
              multiline={true}
              numberOfLines={4}
              onChangeText={[Function]}
              placeholder="A brief description"
              placeholderTextColor="#0A0A0A80"
              selectTextOnFocus={true}
              style={
                Array [
                  Object {
                    "height": 133.5,
                    "lineHeight": 0,
                    "paddingLeft": 24,
                    "width": "100%",
                  },
                ]
              }
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>,
  <View
    style={
      Object {
        "paddingBottom": 48,
        "paddingHorizontal": 48,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#032e90",
          "borderRadius": 20,
          "elevation": 1,
          "height": 80,
          "justifyContent": "center",
          "opacity": 1,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 2,
          "width": "100%",
        }
      }
    >
      <View>
        <Text
          allowFontScaling={false}
          style={
            Object {
              "color": "#e4e8e5",
              "fontFamily": "Montserrat-Bold",
              "fontSize": 27,
            }
          }
        >
          Submit
        </Text>
      </View>
    </View>
  </View>,
]
`;
